### PR TITLE
SYCL: Improve print_configuration

### PR DESCRIPTION
**Pull request by masterleinad**
Related to https://github.com/kokkos/kokkos/pull/6793#discussion_r1483265300, #6758, and #5419.

With this pull request we also print whether we are using `oneDPL` and print all (usable) devices trying to emulate the output from `sycl-ls`, also see https://github.com/kokkos/kokkos/pull/6758#issuecomment-1924433536.

Example (non-verbose):
```
Devices:
  KOKKOS_ENABLE_SYCL: yes

Runtime Configuration:
macro  KOKKOS_ENABLE_SYCL : defined
macro  KOKKOS_IMPL_SYCL_DEVICE_GLOBAL_SUPPORTED : defined
macro  SYCL_EXT_ONEAPI_DEVICE_GLOBAL : defined
macro  KOKKOS_IMPL_SYCL_USE_IN_ORDER_QUEUES : defined
macro  KOKKOS_ENABLE_ONEDPL : defined

Available devices: 
[ext_oneapi_level_zero]:gpu:0] Intel(R) Data Center GPU Max 1550 : Selected
[ext_oneapi_level_zero]:gpu:1] Intel(R) Data Center GPU Max 1550
[ext_oneapi_level_zero]:gpu:2] Intel(R) Data Center GPU Max 1550
[ext_oneapi_level_zero]:gpu:3] Intel(R) Data Center GPU Max 1550
[ext_oneapi_level_zero]:gpu:4] Intel(R) Data Center GPU Max 1550
[ext_oneapi_level_zero]:gpu:5] Intel(R) Data Center GPU Max 1550
```

### CODE REVIEW

File: core/src/SYCL/Kokkos_SYCL.cpp [hunk 1]
```diff
@@ -111,8 +111,28 @@ void SYCL::print_configuration(std::ostream& os, bool verbose) const {
...
+#else
+  os << "macro  KOKKOS_ENABLE_ONEDPL : undefined\n";
+#endif
+
+  int counter       = 0;
+  int active_device = Kokkos::device_id();
+  std::cout << "\nAvailable devices: \n";
+  std::vector<sycl::device> devices = Impl::get_sycl_devices();
+  for (const auto& device : devices) {
+    os << "[" << device.get_backend() << "]:gpu:" << counter << "] "
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL.cpp:125]: Are you sure you always want to print "gpu"?
What about `device->is_gpu() : "gpu" ? ...`

Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL.cpp:125]: `get_sycl_devices` only ever returns GPU devices so the output is consistent at least.

**Comment by masterleinad** [comment]
Now printing
```
Available devices: 
[opencl]:gpu:0] Intel(R) Data Center GPU Max 1550 : Selected
[opencl]:gpu:1] Intel(R) Data Center GPU Max 1550
[opencl]:gpu:2] Intel(R) Data Center GPU Max 1550
[opencl]:gpu:3] Intel(R) Data Center GPU Max 1550
[opencl]:gpu:4] Intel(R) Data Center GPU Max 1550
[opencl]:gpu:5] Intel(R) Data Center GPU Max 1550
[opencl]:cpu:6] Intel(R) Xeon(R) CPU Max 9470C
[ext_oneapi_level_zero]:gpu:7] Intel(R) Data Center GPU Max 1550
[ext_oneapi_level_zero]:gpu:8] Intel(R) Data Center GPU Max 1550
[ext_oneapi_level_zero]:gpu:9] Intel(R) Data Center GPU Max 1550
[ext_oneapi_level_zero]:gpu:10] Intel(R) Data Center GPU Max 1550
[ext_oneapi_level_zero]:gpu:11] Intel(R) Data Center GPU Max 1550
[ext_oneapi_level_zero]:gpu:12] Intel(R) Data Center GPU Max 1550
```
with device filters unset and 
```diff
diff --git a/core/src/SYCL/Kokkos_SYCL.cpp b/core/src/SYCL/Kokkos_SYCL.cpp
index e78d030fe..7feee391c 100644
--- a/core/src/SYCL/Kokkos_SYCL.cpp
+++ b/core/src/SYCL/Kokkos_SYCL.cpp
@@ -271,7 +271,9 @@ std::ostream& SYCL::impl_sycl_info(std::ostream& os,
 namespace Impl {
 
 std::vector<sycl::device> get_sycl_devices() {
-  std::vector<sycl::device> gpu_devices =
+       #if defined(KOKKOS_ARCH_INTEL_GPU) || defined(KOKKOS_IMPL_ARCH_NVIDIA_GPU) || \
+    defined(KOKKOS_ARCH_AMD_GPU)
+  std::vector<sycl::device> devices =
       sycl::device::get_devices(sycl::info::device_type::gpu);
 #if defined(KOKKOS_ARCH_INTEL_GPU) || defined(KOKKOS_IMPL_ARCH_NVIDIA_GPU) || \
     defined(KOKKOS_ARCH_AMD_GPU)
@@ -288,7 +290,10 @@ std::vector<sycl::device> get_sycl_devices() {
                                    }),
                     gpu_devices.end());
 #endif
-  return gpu_devices;
+#else
+   std::vector<sycl::device> devices = sycl::device::get_devices();
+#endif
+  return devices;
 }
 
 int g_sycl_space_factory_initialized =
```

**Review by dalg24 — APPROVED** [review]
I would have reordered
```
macro  KOKKOS_ENABLE_SYCL : defined
macro  KOKKOS_ENABLE_ONEDPL : defined
macro  KOKKOS_IMPL_SYCL_DEVICE_GLOBAL_SUPPORTED : defined
macro  KOKKOS_IMPL_SYCL_USE_IN_ORDER_QUEUES : defined
macro  SYCL_EXT_ONEAPI_DEVICE_GLOBAL : defined
```
or so.
I don't think saying 
that `macro  KOKKOS_ENABLE_SYCL : defined` is useful but I know other backends do that nonsense too

**Review by Rombur — APPROVED** [review]

**Comment by dalg24** [comment]
Ignoring HIP failures

**Pull request merged**